lib/regular_expression/ast/to_nfa.rb

```diff
@@ -0,0 +1,215 @@
+          when Period
+            transition = NFA::Transition::Any.new(finish)
+            start.add_transition(transition)
+          when PositiveLookahead
+            start.add_transition(NFA::Transition::PositiveLookahead.new(finish, node.value))
+          when NegativeLookahead
+            start.add_transition(NFA::Transition::NegativeLookahead.new(finish, node.value))
+          when CharacterRange
+            transition = NFA::Transition::Range.new(finish, node.left, node.right)
+            start.add_transition(transition)
+          when Anchor
+            transition =
+              case node.value
+              when %q{\A}
+                NFA::Transition::BeginAnchor.new(finish)
+              when %q{\z}, %q{$}
+                NFA::Transition::EndAnchor.new(finish)
+              end
+
+            start.add_transition(transition)
+          when AddEpsilon
+            start.add_transition(NFA::Transition::Epsilon.new(finish))
+          else
+            raise "Unexpected work item: #{node.inspect}"
+          end
+        end
+
+        @start_state
+      end
+
+      private
+
+      def pop_work
+        @worklist.pop
+      end
+
+      def push_work(work)
+        work.reverse_each { |work_item| @worklist << work_item }
+      end
+
+      def quantify(quantifier, start, finish)
+        work = []
+
+        case quantifier
+        when Quantifier::Once
+          yield start, finish, work
+        when Quantifier::ZeroOrMore
+          yield start, start, work
+          work << [AddEpsilon, start, finish]
+        when Quantifier::OneOrMore
+          yield start, finish, work
+          work << [AddEpsilon, finish, start]
+        when Quantifier::Optional
+          yield start, finish, work
+          work << [AddEpsilon, start, finish]
+        when Quantifier::Exact
+          states = [start, *(quantifier.value - 1).times.map { NFA::State.new(@labels.next) }, finish]
+
+          quantifier.value.times do |index|
+            yield states[index], states[index + 1], work
+          end
+        when Quantifier::AtLeast
+          states = [start, *(quantifier.value - 1).times.map { NFA::State.new(@labels.next) }, finish]
+
+          quantifier.value.times do |index|
+            yield states[index], states[index + 1], work
+          end
+
+          work << [AddEpsilon, states[-1], states[-2]]
+        when Quantifier::Range
+          states = [start, *(quantifier.upper - 1).times.map { NFA::State.new(@labels.next) }, finish]
+
+          quantifier.upper.times do |index|
+            yield states[index], states[index + 1], work
+          end
+
+          (quantifier.upper - quantifier.lower).times do |index|
+            work << [AddEpsilon, states[quantifier.lower + index], states[-1]]
+          end
+        else
+          raise quantify
+        end
+
+        push_work(work)
+      end
+
+      AddEpsilon = Object.new
+      private_constant(:AddEpsilon)
```

Comment:
We are introducing this `AddEpsilon` as a way of delaying the work of pushing the epsilon transition to the state's `transitions` array. Delaying the work is needed in order to keep adding the epsilons last, otherwise our generated bytecode won't be the same.

This is a bit of a hack -- all other work items are `[AST node, start, finish]`. With this one, we're introducing a non AST-node in position 0. But 🤷  best we came up with. Ideas?